helper/ldaputil/client.go

```diff
@@ -0,0 +1,366 @@
+package ldaputil
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"math"
+	"net"
+	"net/url"
+	"strings"
+	"text/template"
+
+	"github.com/go-ldap/ldap"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/helper/tlsutil"
+)
+
+type Client struct {
+	Logger hclog.Logger
+	LDAP   LDAP
+}
```

Comment:
I made the design decision to allow the logger to be passed in here and to persist, but for the config to always be passed in to methods rather than persisted on the client. This is because if there's an updated config at any time, I don't want a possibly long-lived client to be holding onto an old config forever, causing Vault to have to be restarted for it to be picked up. I realize that people could work around that by making new clients all the time, but I thought that seemed like a PIA.

Reply:
👍